packages/commonwealth/server/util/cosmosCache.ts

```diff
@@ -0,0 +1,113 @@
+import type { Request, RequestHandler } from 'express';
+import { cacheDecorator } from 'common-common/src/cacheDecorator';
+
+export const cosmosCacheRPC = (duration: number): RequestHandler => {
+  const cacheFunc = (ttl: number) =>
+    cacheDecorator.cacheMiddleware(ttl, cosmosRPCKeyGenerator);
+
+  // RPC specific codes from cosmJS requests
+  const activeCodes = ['0801', '0802'];
+  const completedCodes = [
+    '0803',
+    '0804',
+    '0805',
+    '0803220a0a080000000000000087',
+    '0803220a0a080000000000000100',
+    '0803220a0a080000000000000189',
+  ];
+
+  return async function cacheIfCacheworthyRPC(req, res, next) {
+    let body = req.body;
+    try {
+      body = JSON.parse(req.body);
+    } catch (err) {
+      console.log('Error parsing body: ', err);
+    }
+
+    const params = JSON.stringify(body?.params);
+    const method = body?.method;
+
+    if (/^(status|block|tx)/.test(method)) {
+      // LIVENESS CHECK  or TX Response: do not cache
+      return next();
+    }
+
+    if (/^Query\/(Votes|TallyResult|Pool|Deposits)/.test(params)) {
+      return next(); // LIVE data matters do not cache
+    }
+
+    if (activeCodes.some((code) => code === body?.params?.data)) {
+      // ACTIVE PROPOSALS: do not cache
+      return next();
+    }
+
+    if (params.includes('Query/Params')) {
+      // rpc PARAMS: cache long-term (1 day)
+      return cacheFunc(60 * 60 * 24)(req, res, next);
+    }
+
+    if (completedCodes.some((code) => code === body?.params?.data)) {
+      // COMPLETED PROPOSALS: cache medium-term (30 min)
+      return cacheFunc(60 * 30)(req, res, next);
+    }
```

Comment:
TTL could be shorter here. Because we don't want recently-completed proposals to have a lag before showing up in "completed" section. Maybe 60 seconds.